test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,316 @@
+)
+
+func TestFQDNPolicyWithCachedDNS(t *testing.T) {
+	skipIfAntreaPolicyDisabled(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	customDnsService, err := data.CreateServiceWithAnnotations(customDnsServiceName, data.testNamespace, customDnsPort, customDnsTargetPort, corev1.ProtocolUDP,
+		map[string]string{customDnsLabelKey: customDnsLabelValue}, false, false, corev1.ServiceTypeClusterIP, ptr.To[corev1.IPFamily](corev1.IPv4Protocol), map[string]string{})
+	require.NoError(t, err, "Error when creating custom DNS service: %v", err)
+
+	setDnsServerAddressInAntrea(t, data, customDnsService.Spec.ClusterIP)
+
+	var agnHostPodIps []*PodIPs
+	podCount := 2
+	for i := 0; i < podCount; i++ {
+		podIp := createHttpAgnhostPod(t, data, randName(agnHostPodNamePreFix), map[string]string{agnHostLabelKey: agnHostLabelValue})
+		agnHostPodIps = append(agnHostPodIps, podIp)
+	}
+	//domainMapping holds whether the IP is mapped to Domain or not.
+	domainMapping := make(map[string]bool)
+
+	// pick an IP to be added in config
+	var ipForDnsConfig string
+	for idx, podIp := range agnHostPodIps {
+		ipStr := podIp.IPv4.String()
+		domainMapping[ipStr] = false
+		//pick last IP for config
+		if idx == len(agnHostPodIps)-1 {
+			ipForDnsConfig = ipStr
+		}
+	}
+
+	dnsConfigData := createDnsConfig(ipForDnsConfig)
+	customDnsConfigMapObject, err := data.CreateConfigMap(data.testNamespace, customDnsConfigName, dnsConfigData, nil, false)
+	require.NoError(t, err, "failed to create custom dns ConfigMap: %v", err)
+	domainMapping[ipForDnsConfig] = true
+
+	createDnsPod(t, data)
+
+	k8sUtils, err = NewKubernetesUtils(data)
+	require.NoError(t, err, "error getting k8s utils %+v", err)
+
+	_ = buildFqdnPolicy(t, data)
+
+	createToolBoxPod(t, data, customDnsService.Spec.ClusterIP)
+
+	curlFqdn := func(podName, containerName, fqdn string) error {
+		cmd := []string{"curl", fqdn}
+		stdout, stderr, err := data.RunCommandFromPod(data.testNamespace, podName, containerName, cmd)
+		if err != nil {
+			return fmt.Errorf("error when running command '%s' on Pod '%s': %v, stdout: <%v>, stderr: <%v>",
+				strings.Join(cmd, " "), podName, err, stdout, stderr)
+		}
+		return nil
+	}
+
+	assert.Eventually(t, func() bool {
+		t.Logf("trying to curl the fqdn %v", testFullyQualifiedDomainName)
+		err = curlFqdn(toolBoxPodName, toolboxContainerName, testFullyQualifiedDomainName)
+		return assert.NoError(t, err)
+	}, 2*time.Second, 100*time.Millisecond)
+
+	// DIG to get actual IP , to be sure.
+	digResponse, err := k8sUtils.runDNSQuery(toolBoxPodName, toolboxContainerName, data.testNamespace, testFullyQualifiedDomainName, false, customDnsService.Spec.ClusterIP)
+	require.NoError(t, err, "failed to get IP of FQDN using DIG from toolbox pod : %v", err)
+	fqdnIp := digResponse.String()
+
+	t.Logf("received ip using dig for test fqdn %+v ", fqdnIp)
+
+	var newIP string
+	for ip, mapped := range domainMapping {
```

Comment:
I have a lot of trouble understand what the `domainMapping` map / set is for. You already know the old IP(s) and the new IP(s) (first Pod / second Pod). You can have variables for these IPs, and there is no need to store them in a map.